stan/math/prim/meta/is_eigen_matrix_or_array.hpp

```diff
@@ -0,0 +1,28 @@
+#ifndef STAN_MATH_PRIM_META_IS_EIGEN_MATRIX_OR_ARRAY_HPP
+#define STAN_MATH_PRIM_META_IS_EIGEN_MATRIX_OR_ARRAY_HPP
+
+#include <stan/math/prim/fun/Eigen.hpp>
+#include <stan/math/prim/meta/bool_constant.hpp>
+#include <stan/math/prim/meta/disjunction.hpp>
+#include <stan/math/prim/meta/is_eigen_matrix.hpp>
+#include <stan/math/prim/meta/is_eigen_array.hpp>
+#include <type_traits>
+
+namespace stan {
+
+/** \addtogroup type_trait
+ *  @{
+ */
+
+/**
+ * Check if a type satisfies either of @c is_eigen_matrix or @c is_eigen_array
+ * @tparam T Type to check if it is derived from `EigenBase`
+ */
+template <typename T>
+using is_eigen_matrix_or_array
+    = math::disjunction<is_eigen_matrix<std::decay_t<T>>,
+                        is_eigen_array<std::decay_t<T>>>;
```

Comment:
With the change to `is_eigen_matrix` this does not make sense anymore. It should be renamed to `is_eigen_matrix_or_vector_or_array` ant the test should be changed appropriately.

Reply:
It seems like the only place we are using `is_eigen_matrix_or_array` was in matrix_cl to check whether we should call .eval() or not. `is_eigen_matrix_or_vector_or_array` seems like kind of a mouthful. Should we just do `is_eigen_dense` in `matrix_cl` and get rid of this type trait?

Code that uses `is_eigen_matrix_or_array:
https://github.com/stan-dev/math/blob/develop/stan/math/opencl/matrix_cl.hpp#L312

Reply:
`is_eigen_dense` is not the same thing! But I agree we can find a better name. This is actually checking whether something is a plain type so maybe `is_eigen_plain_type`?